vsp.go

```diff
@@ -110,6 +114,240 @@ func (v *VSP) PurchaseTickets(ticketCount, expiryBlocks int32, passphrase []byte
+		ctx, cancel := context.WithCancel(v.ctx)
+		v.w.cancelAutoTicketBuyer = cancel
+
+		err := v.ticketBuyerInstance(ctx, balanceToMaintain, passphrase)
+		if err != nil {
+			log.Errorf("Ticket buyer instance errored: %v", err)
+		}
+
+		v.w.cancelAccountMixer = nil
+	}()
+
+	return nil
+}
+
+func (v *VSP) ticketBuyerInstance(ctx context.Context, balanceToMaintain int64, passphrase []byte) error {
+	if len(passphrase) > 0 {
+		err := v.w.UnlockWallet(passphrase)
+		if err != nil {
+			return translateError(err)
+		}
+	}
+
+	c := v.w.Internal().NtfnServer.MainTipChangedNotifications()
+	defer c.Done()
+
+	ctx, outerCancel := context.WithCancel(ctx)
+	defer outerCancel()
+	var fatal error
+	var fatalMu sync.Mutex
+
+	var nextIntervalStart, expiry int32
+	var cancels []func()
+	for {
+		select {
+		case <-ctx.Done():
+			defer outerCancel()
+			fatalMu.Lock()
+			err := fatal
+			fatalMu.Unlock()
+			if err != nil {
+				return err
+			}
+			return ctx.Err()
+		case n := <-c.C:
+			if len(n.AttachedBlocks) == 0 {
+				continue
+			}
+
+			tip := n.AttachedBlocks[len(n.AttachedBlocks)-1]
+			w := v.w.Internal()
+
+			// Don't perform any actions while transactions are not synced through
+			// the tip block.
+			rp, err := w.RescanPoint(ctx)
+			if err != nil {
+				return err
+			}
+			if rp != nil {
+				log.Info("Skipping autobuyer actions: transactions are not synced")
+				continue
+			}
+
+			tipHeader, err := w.BlockHeader(ctx, tip)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			height := int32(tipHeader.Height)
+
+			// Cancel any ongoing ticket purchases which are buying
+			// at an old ticket price or are no longer able to
+			// create mined tickets the window.
+			if height+2 >= nextIntervalStart {
+				for i, cancel := range cancels {
+					cancel()
+					cancels[i] = nil
+				}
+				cancels = cancels[:0]
+
+				intervalSize := int32(w.ChainParams().StakeDiffWindowSize)
+				currentInterval := height / intervalSize
+				nextIntervalStart = (currentInterval + 1) * intervalSize
+
+				// Skip this purchase when no more tickets may be purchased in the interval and
+				// the next sdiff is unknown.  The earliest any ticket may be mined is two
+				// blocks from now, with the next block containing the split transaction
+				// that the ticket purchase spends.
+				if height+2 == nextIntervalStart {
+					log.Info("Skipping purchase: next sdiff interval starts soon")
+					continue
+				}
+				// Set expiry to prevent tickets from being mined in the next
+				// sdiff interval.  When the next block begins the new interval,
+				// the ticket is being purchased for the next interval; therefore
+				// increment expiry by a full sdiff window size to prevent it
+				// being mined in the interval after the next.
+				expiry = nextIntervalStart
+				if height+1 == nextIntervalStart {
+					expiry += intervalSize
+				}
+			}
+
+			cancelCtx, cancel := context.WithCancel(ctx)
+			cancels = append(cancels, cancel)
+			go func() {
```

Comment:
I notice here that dcrwallet appears to have support for buying more than 1 ticket per block, but we don't have that here. I also see that the dcrwallet ticket buyer has some relationship with the mixer but I haven't dived deep to see the nature of the relationship. I did notice though that dcrlibwallet's account mixer seems to set up a ticket buyer as well and wonder why we aren't using that here.

https://github.com/planetdecred/dcrlibwallet/blob/1cdba274e194b8c8090b82026bfbf970e4668133/account_mixer.go#L164
https://github.com/planetdecred/dcrlibwallet/blob/1cdba274e194b8c8090b82026bfbf970e4668133/account_mixer.go#L200-L208
https://github.com/planetdecred/dcrlibwallet/blob/1cdba274e194b8c8090b82026bfbf970e4668133/account_mixer.go#L223

Reply:
The limits to the number of tickets to be purchase is currently set to `TODO` the feature to increase the limits per blocks will be added later on. It is currently set to default `1` due to the meeting had with raedah some few day ago.

Ticket mixer and ticket buyer for `dcrwallet` share same functionally but runs different process depending on the flag set. `c.MixChange` and `c.BuyTickets`.
The reason i am not using the dcrwallet ticket buyer dirrectly is the `c.VSP` config param. it is not letting me set the value to dcrlibwallet `vspClient`.

Reply:
Interesting point there. I have a diff in the works that may allow you set the c.VSP param. Will update later.